### ISSUE TITLE
Correct implementation of the Loss function

### ISSUE THREAD

**Issue by chinmay5** [ZihaoWang-CV/CAMP_iccv19#2]
Hi trying to implement the code and am having difficulty in the implementation of the loss function.

This is my code :-

`    
   # Function for computing the loss
     def forward(self, v_final, t_final, model):
        eps = 1e-7
        scores = self.element_wise_sum(v_final, t_final)
        batch_size = scores.shape[0]
        scores_tensor = scores.reshape(-1, scores.shape[2])
        output = model.TwoLayerMLP(scores_tensor)
        scores = output.reshape(batch_size, batch_size)
        probs = torch.sigmoid(scores)

        # match non-diagonals
        mask = torch.eye(scores.size(0)) < .5  # Only diagonal entries would be false
        mask = mask.cuda()
        neg_probs = probs.masked_select(mask)
        neg_probs = torch.log((1 - neg_probs) + eps)
        neg_probs = neg_probs.reshape(batch_size, -1)

        cost_s = neg_probs.max(1)[0]
        cost_im = neg_probs.max(0)[0]

        # Now take the diagonal entries as
        pos_loss = probs.masked_select(torch.eye(batch_size, dtype=torch.bool).cuda())
        pos_loss = torch.log(pos_loss + eps)

        # Now let us sum all the values
        loss = (cost_im.sum() + cost_s.sum() + pos_loss.sum()/scores.shape[0])
        return - loss

    @staticmethod
    def element_wise_sum(text, image):
        distance_matrix = text + image.unsqueeze(1)
        return distance_matrix
`
It's mostly the selection of hard negative samples that is confusing me. Your inputs about the same
would be really really helpful

**Comment by ZihaoWang-CV** [comment]
Thanks for your interest.
Sorry for the delayed release, I have left Sensetime and the opensource request has not been approved yet. But I think I could partly share the class for the loss function here.
`

    Class SimLoss(nn.Module):
    def __init__(self, margin=0, measure=False, max_violation=False, inner_dim=0, loss_func="BCE"):
        super(SimLoss, self).__init__()
        self.margin = margin
        self.measure = measure
        if measure == 'cosine':
            self.sim = cosine_sim
        elif measure == 'attention':
            self.sim = attention_sim
        elif measure == 'cross_attention':
            self.sim = CrossAttention(inner_dim, 4, -1)
        elif measure == 'cross_attention_new':
            self.sim = CrossAttentionNew(inner_dim, 4, -1)
        elif measure == 'gate_fusion':
            self.sim = GatedFusion(inner_dim, 4, 0.0)
        elif measure == 'gate_fusion_new':
            self.sim = GatedFusionNew(inner_dim, 4, 0.0)
        else:
            self.sim = cosine_sim

        self.loss_func = loss_func
        self.max_violation = max_violation

    def forward(self, im, s, get_score=False, keep="words", mask=None):
        # compute image-sentence score matrix
        if self.measure == 'cosine':
            cur_im = im
            cur_s = s
            drive_num = torch.cuda.device_count()

            if keep == "words":
                cur_s = s.unsqueeze(0).expand(drive_num, -1, -1, -1)
            elif keep == "regions":
                cur_im = im.unsqueeze(0).expand(drive_num, -1, -1, -1)

            scores = self.sim(cur_im, cur_s, keep=keep, ret_dot=True)
            
            if keep == "regions":
                scores = scores.transpose(0, 1)
        elif self.measure == 'cross_attention' or self.measure == 'cross_attention_new':
            cur_im = im
            cur_s = s
            cur_mask = mask
            drive_num = torch.cuda.device_count()

            if keep == "words":
                cur_s = s.unsqueeze(0).expand(drive_num, -1, -1, -1)
                cur_mask = mask.unsqueeze(0).expand(drive_num, -1, -1)
            elif keep == "regions":
                cur_im = im.unsqueeze(0).expand(drive_num, -1, -1, -1)

            scores = self.sim(cur_im, cur_s, keep=keep, mask=cur_mask)
            
            if keep == "regions":
                scores = scores.transpose(0, 1)
        elif self.measure == 'gate_fusion' or self.measure == 'gate_fusion_new':
            cur_im = im
            #print("cur_im:", cur_im)
            cur_s = s
            cur_mask = mask
            drive_num = torch.cuda.device_count()

            if keep == "words":
                cur_s = s.unsqueeze(0).expand(min(im.size(0), drive_num), -1, -1, -1)
                cur_mask = mask.unsqueeze(0).expand(min(im.size(0), drive_num), -1, -1)
            elif keep == "regions":
                cur_im = im.unsqueeze(0).expand(drive_num, -1, -1, -1)

            scores = self.sim(cur_im, cur_s, keep=keep, mask=cur_mask)

            if keep == "regions":
                scores = scores.transpose(0, 1)
        else:
            scores = self.sim(im, s)

        if get_score:
            return scores


        if self.loss_func == 'BCE':
            eps = 0.000001

            scores = scores.clamp(min=eps, max=(1.0-eps))
            de_scores = 1.0 - scores

            label = Variable(torch.eye(scores.size(0))).cuda()
            de_label = 1 - label
        
            scores = torch.log(scores) * label
            de_scores = torch.log(de_scores) * de_label

            if self.max_violation:
                le = -(scores.sum() + scores.sum() + de_scores.min(1)[0].sum() + de_scores.min(0)[0].sum())
            else:
                le = -(scores.diag().mean() + de_scores.mean())

            return le
        else:
            
            diagonal = scores.diag().view(im.size(0), 1)
            d1 = diagonal.expand_as(scores)
            d2 = diagonal.t().expand_as(scores)
            # compare every diagonal score to scores in its column
            # caption retrieval
            cost_s = (self.margin + scores - d1).clamp(min=0)
            # compare every diagonal score to scores in its row
            # image retrieval
            cost_im = (self.margin + scores - d2).clamp(min=0)
            # clear diagonals
            mask = torch.eye(scores.size(0)) > .5
            I = Variable(mask)
            if torch.cuda.is_available():
                I = I.cuda()
            cost_s = cost_s.masked_fill_(I, 0)
            cost_im = cost_im.masked_fill_(I, 0)

            # keep the maximum violating negative for each query
            if self.max_violation:
                cost_s = cost_s.max(1)[0]
                cost_im = cost_im.max(0)[0]

            return cost_s.sum() + cost_im.sum()
`


For distributed training, we would use `nn.DataParallel` to create the class:
`

    self.criterion.sim = nn.DataParallel(self.criterion.sim)
    self.criterion.sim.cuda()
`

This is a trick to compute the similarity matrix in one machine with multiple cards. 
If you want to compute an `**n*n** `score matrix with the image tensor with shape `(n, region_num, feature_dim)` and the sentence tensor with shape `(n, word_num, feature_dim)` on `device_num` GPUs, the class will compute a similarity matrix with shape `n*(n/device_num)` on each GPU and gather those matrices to concatenate the final similarity matrix.

For selecting the hardest negative examples, we follow the implementation of VSE++, choose the max value for each row and line of the score matrix.
